iOS/Flows/Conversation/Cells/EmotionCircleCollectionViewLayout.swift

```diff
@@ -0,0 +1,187 @@
+    private let itemBehavior = UIDynamicItemBehavior()
+    private let noiseField = UIFieldBehavior.noiseField(smoothness: 1, animationSpeed: 0.01)
+
+    private let cellDiameter: CGFloat = 100
+
+    override init() {
+        super.init()
+
+        self.collisionBehavior.translatesReferenceBoundsIntoBoundary = true
+        self.collisionBehavior.collisionMode = .boundaries
+        self.animator.addBehavior(self.collisionBehavior)
+
+        self.itemBehavior.elasticity = 1
+        self.itemBehavior.friction = 0
+        self.itemBehavior.resistance = 0
+        self.itemBehavior.angularResistance = 0
+        self.animator.addBehavior(self.itemBehavior)
+
+        self.noiseField.strength = 10
+        self.animator.addBehavior(self.noiseField)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override var collectionViewContentSize: CGSize {
+        // The collection view is not scrollable.
+        return self.collectionView?.bounds.size ?? .zero
+    }
+
+    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
+        guard let collectionView = self.collectionView else { return false }
+
+        return newBounds.size != collectionView.frame.size
+    }
+
+    override func prepare() {
+        super.prepare()
+
+        guard let collectionView = self.collectionView else { return }
+
+        // Get all of the emotion attributes being managed by the animator.
+        let animatorAttributes
+        = self.animator.items(in: collectionView.bounds).compactMap { item in
+            return item as? EmotionCircleAttributes
+        }
+
+        // Keep track of the ids of items that currently exist in the data source.
+        var foundIds = Set<String>()
+
+        self.forEachIndexPath { indexPath in
+            // Get the id of the attributes and save it for later use.
+            let id = self.dataSource.getId(forItemAt: indexPath)
+
+            foundIds.insert(id)
+            // Check to see if the attribute already exists in the animator
+            if let matchingAttributes = animatorAttributes.first(where: { attributes in
+                return attributes.id == id
+            }) {
+                // The attributes were already added, but update the index path in case
+                // it's no longer valid due to inserts/deletes.
+                matchingAttributes.indexPath = indexPath
+            } else {
+                // The item's attributes aren't yet being managed by the animator. Add them now.
+                self.addEmotionAttributesToAnimator(for: indexPath, withId: id)
+            }
+        }
+
+        // Removing any attributes from the animator that no longer exists.
+        animatorAttributes.forEach { attributes in
+            guard !foundIds.contains(attributes.id) else { return }
+            self.removeDynamicItemFromAnimator(attributes)
+        }
```

Comment:
When dealing with a dynamic animator, prepare gets called every frame. This is because the layout is constantly being invalidated as the physics engine moves the cells around.

This code makes sure we clean up layout attributes for cells that were deleted and add new ones.